packages/arb-avm-cpp/data_storage/src/value/valuecache.cpp

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2019-2020, Offchain Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <data_storage/value/valuecache.hpp>
+
+void ValueCache::clear() {
+    cache.clear();
+}
+
+void ValueCache::maybeSave(value val) {
+    cache.emplace(hash_value(val), std::move(val));
```

Comment:
FYI this is safe, but this code is a bit risky since evaluation order of function arguments is unspecified https://en.cppreference.com/w/cpp/language/eval_order